src/query/expression/src/deserializations/decimal.rs

```diff
@@ -0,0 +1,373 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::cmp::Ordering;
+
+use common_exception::ErrorCode;
+use common_exception::Result;
+use common_io::prelude::FormatSettings;
+
+use crate::types::decimal::Decimal;
+use crate::types::decimal::DecimalSize;
+use crate::types::DecimalDataType;
+use crate::Column;
+use crate::Scalar;
+use crate::TypeDeserializer;
+
+pub struct DecimalDeserializer<T: Decimal> {
+    pub values: Vec<T>,
+    pub ty: DecimalDataType,
+    // for fast access
+    pub size: DecimalSize,
+}
+
+impl<T: Decimal> DecimalDeserializer<T> {
+    pub fn with_capacity(ty: &DecimalDataType, capacity: usize) -> Self {
+        Self {
+            size: ty.size(),
+            ty: *ty,
+            values: Vec::with_capacity(capacity),
+        }
+    }
+}
+
+impl<T: Decimal> DecimalDeserializer<T> {
+    pub fn de_json_inner(&mut self, value: &serde_json::Value) -> Result<()> {
+        match value {
+            serde_json::Value::Number(n) => {
+                if n.is_i64() {
+                    self.values.push(T::from_i64(n.as_i64().unwrap()));
+                    Ok(())
+                } else if n.is_u64() {
+                    self.values.push(T::from_u64(n.as_u64().unwrap()));
+                    Ok(())
+                } else {
+                    Err(parse_error("Incorrect json value for decimal"))
+                }
+            }
+            serde_json::Value::String(s) => {
+                let (n, _) = read_decimal_with_size::<T>(s.as_bytes(), self.size, true)?;
+                self.values.push(n);
+                Ok(())
+            }
+            _ => Err(ErrorCode::from("Incorrect json value for decimal")),
+        }
+    }
+}
+
+impl<T: Decimal> TypeDeserializer for DecimalDeserializer<T> {
+    fn memory_size(&self) -> usize {
+        self.values.len() * T::mem_size()
+    }
+
+    fn len(&self) -> usize {
+        self.values.len()
+    }
+
+    // See GroupHash.rs for StringColumn
+    #[allow(clippy::uninit_vec)]
+    fn de_binary(&mut self, _reader: &mut &[u8], _format: &FormatSettings) -> Result<()> {
+        todo!()
+    }
+
+    fn de_default(&mut self) {
+        self.values.push(T::zero());
+    }
+
+    fn de_fixed_binary_batch(
+        &mut self,
+        _reader: &[u8],
+        _step: usize,
+        _rows: usize,
+        _format: &FormatSettings,
+    ) -> Result<()> {
+        todo!()
+    }
+
+    fn de_json(&mut self, value: &serde_json::Value, _format: &FormatSettings) -> Result<()> {
+        match value {
+            serde_json::Value::Number(n) => {
+                if n.is_i64() {
```

Comment:
Seems it's wrong, we should care about the `precision` ? See `m_integer_to_decimal`

Reply:
@sundy-li  https://github.com/datafuselabs/databend/pull/10167/files#diff-93bbc5f6f54e7d2224ea2ff14e7dcf663af5d93c401dab999039c144f3cbf8d9R238

Reply:
Ok, better to add a test to cover this bug.